pjlib/src/pj/ssl_sock_apple.m

```diff
@@ -0,0 +1,2013 @@
+
+    parameters = nw_parameters_create_secure_tcp(configure_tls,
+    		     NW_PARAMETERS_DEFAULT_CONFIGURATION);
+
+    protocol_stack = nw_parameters_copy_default_protocol_stack(parameters);
+    ip_options = nw_protocol_stack_copy_internet_protocol(protocol_stack);
+    if (ssock->param.sock_af == pj_AF_INET()) {
+	nw_ip_options_set_version(ip_options, nw_ip_version_4);
+    } else if (ssock->param.sock_af == pj_AF_INET6()) {
+	nw_ip_options_set_version(ip_options, nw_ip_version_6);
+    }
+    nw_release(ip_options);
+    nw_release(protocol_stack);
+    
+    if (ssock->is_server && ssock->param.reuse_addr) {
+    	nw_parameters_set_reuse_local_address(parameters, true);
+    }
+
+    /* Create local endpoint.
+     * Currently we ignore QoS and socket options.
+     */
+    pj_sockaddr_print(localaddr, ip_addr,sizeof(ip_addr),0);
+
+    if (port_range) {
+	pj_uint16_t max_try = MAX_BIND_RETRY;
+
+	if (port_range && port_range < max_try) {
+	    max_try = port_range;
+	}
+	for (; max_try; --max_try) {
+	    pj_uint16_t base_port;
+
+	    base_port = pj_sockaddr_get_port(localaddr);
+	    port = (pj_uint16_t)(base_port + pj_rand() % (port_range + 1));
+	    pj_utoa(port, port_str);
+	    
+	    local_endpoint = nw_endpoint_create_host(ip_addr, port_str);
+	    if (local_endpoint)
+	    	break;
+	}
+    } else {
+    	port = pj_sockaddr_get_port(localaddr);
+    	pj_utoa(port, port_str);
+
+    	local_endpoint = nw_endpoint_create_host(ip_addr, port_str);
+    }
+    
+    if (!local_endpoint) {
+        PJ_LOG(2, (THIS_FILE, "Failed creating local endpoint"));
+    	return PJ_EINVALIDOP;
+    }
+
+    nw_parameters_set_local_endpoint(parameters, local_endpoint);
+    nw_release(local_endpoint);
+    
+    *p_params = parameters;
+    return PJ_SUCCESS;
+}
+
+/* Setup assock's connection state callback and start the connection */
+static pj_status_t network_setup_connection(pj_ssl_sock_t *ssock,
+					    void *connection)
+{
+    applessl_sock_t *assock = (applessl_sock_t *)ssock;
+    assock->connection = (nw_connection_t)connection;
+    pj_status_t status;
+
+    /* Initialize input circular buffer */
+    status = circ_init(ssock->pool->factory, &ssock->circ_buf_input, 8192);
+    if (status != PJ_SUCCESS)
+        return status;
+
+    /* Initialize output circular buffer */
+    status = circ_init(ssock->pool->factory, &ssock->circ_buf_output, 8192);
+    if (status != PJ_SUCCESS)
+        return status;
+
+    /* Hold a reference until cancelled */
+    nw_retain(assock->connection);
+    nw_connection_set_queue(assock->connection, assock->queue);
+
+    nw_connection_set_state_changed_handler(assock->connection,
+    	^(nw_connection_state_t state, nw_error_t error)
+    {
+    	pj_status_t status = PJ_SUCCESS;
+    	pj_bool_t call_cb = PJ_FALSE;
+
+	if (error) {
+    	    errno = nw_error_get_error_code(error);
+            warn("Connection failed %p", assock);
+            status = PJ_EEOF;
+            call_cb = PJ_TRUE;	
```

Comment:
instead of PJ_EEOF, better to use (PJ mapped) native OS error code if possible, useful for debugging.
also for other network operations, better to use native OS error code, if possible. see also PJ_STATUS_FROM_OS, PJ_STATUS_TO_OS.